basic pwa setup

technophile-04 opened the pull request:
## Description

Basic PWA setup and update the README, lol I think the README is not that great but please feel free to suggest 🙌 

## Additional Information

- [X] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [X] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

## Related Issues

Your ENS/address: shivbhonde.eth

technophile-04 committed:
copy-paste basic pwa

technophile-04 committed:
add .env.example

technophile-04 committed:
minor tweaks to READEME

Pabl0cks commented:
gj Shiv!! 🙌

I'm trying this out to add extra details to Readme with the stuff I find difficult to execute during the setup, and found myself in trouble while trying to run `yarn vercel`. I had to change the import from `contract.ts` to use this route to make it work:

```tsx
import { Prettify } from "viem/types/utils";
```

Think is related with SE2 [issue #419](https://github.com/scaffold-eth/scaffold-eth-2/issues/419), because I found the same problem while trying to yarn vercel a fresh SE2. Here are the details (screenshots from my SE2 test but the errors were the same on this PR):

`yarn vercel`, throws an error while building. Those are details for the errors on Vercel console:

![image](https://github.com/BuidlGuidl/PWA-burner-wallet/assets/55535804/885a8426-b641-408b-8db9-67eac03ee9de)

Error from bash console on Windows:

```bash
$ yarn vercel
Vercel CLI 28.15.1
? Set up and deploy “E:\GitHub\scaffold-eth-2\packages\nextjs”? [Y/n] y
? Set up and deploy “E:\GitHub\scaffold-eth-2\packages\nextjs”? [Y/n] y
Loading scopes…
? Which scope do you want to deploy to?
? Which scope do you want to deploy to? tokodev
Searching for existing projects…
? Link to existing project? [y/N] n
? Link to existing project? [y/N] n
? What’s your project’s name? (nextjs) se2-test-viem-folder
? What’s your project’s name? se2-test-viem-folder
Verifying project name…
? In which directory is your code located? ./
? In which directory is your code located? ./
Setting up project
Uploading [--------------------] (0.0B/80.4KB)
Uploading [======--------------] (24.2KB/80.4KB)
Uploading [=============-------] (55.4KB/80.4KB)
Uploading [=================---] (71.4KB/80.4KB)
Uploading [====================] (80.4KB/80.4KB)
Local settings detected in vercel.json:
Auto-detected Project Settings (Next.js):
- Build Command: next build
- Development Command: next dev --port $PORT
- Install Command: `yarn install`, `pnpm install`, `npm install`, or `bun install`
- Output Directory: Next.js default
? Want to modify these settings? [y/N] n
? Want to modify these settings? [y/N] n
Deploying tokodev/se2-test-viem-folder
Linked to tokodev/se2-test-viem-folder (created .vercel)
Inspect: https://vercel.com/tokodev/se2-test-viem-folder/8n1XJey375nRJ5T2jp9aHy4BPwbH [1s]
https://se2-test-viem-folder-a65w9fbhz-tokodev.vercel.appQueued
Building
Error: Unexpected error. Please try again later. ()
```

Pabl0cks committed:
Update default targetNetwork to hardhat

Pabl0cks committed:
README v0 tweaks

Adding some extra steps and info to install, formatting and re-wording some sections

Pabl0cks commented:
I've just added some additional info to the Readme, and re-worded / formatted some sections. 

We'll have to add all the features explanations when we have them, and keep improving current content in further iterations 🙌

carletex commented:
Great work @technophile-04!! And thanks @Pabl0cks for the review.

I'll give it a go this week. Will also set up the auto-deployment in Vercel & set up a Firebase

technophile-04 committed:
update defaultNetwork to localhost

carletex commented on README.md:
````diff
@@ -1,73 +1,122 @@
 
-```
-git clone https://github.com/scaffold-eth/scaffold-eth-2.git
-cd scaffold-eth-2
-yarn install
+Checkout [this article](https://softauthor.com/add-firebase-to-javascript-web-app/#add-firebase-sdk-to-javascript-web-app) on how to get config variables from Firebase.
+
+> Hint: If it's your first time using Firebase, please read the article from the beginning, as you'll need to create a Firebase project and register a web app first.
+
+We use the Firebase config variables in our `packages/nextjs/database/firebase/config` file:
+
+```js
+// Your web app's Firebase configuration
+const firebaseConfig = {
+  apiKey: process.env.FB_API_KEY,
+  authDomain: process.env.FB_AUTH_DOMAIN,
+  projectId: process.env.FB_PROJECT_ID,
+  storageBucket: process.env.FB_STRG_BUCKET,
+  messagingSenderId: process.env.FB_MESSAGING_SENDER_ID,
+  appId: process.env.FB_APP_ID,
+};
 ```
 
````
Suggestion for this part (I think we can simplify and not sure about linking a weird article :D)

----

Create your [Firebase project](https://console.firebase.google.com/) and register a web app. It'll give you a config object that looks like this:

```js
// Your web app's Firebase configuration
const firebaseConfig = {
  apiKey: "...",
  authDomain: "...",
  projectId: "...",
  storageBucket: "...",
  messagingSenderId: "...",
  appId: "...",
};
```

Copy `packages/nextjs/.env.example` into `packages/nextjs/.env` file and fill in your Firebase credentials which starts with `FIREBASE_`.

```

Pabl0cks replied:
Yep, that section is probably the most confusing part of the setup and has to be simplified and explained better 🙌

I've just pushed the simplified version in this commit: https://github.com/BuidlGuidl/PWA-burner-wallet/pull/1/commits/4af299ac1d8abcd727f7524e546a89b49cb5d9f1

carletex commented on README.md:
```diff
@@ -1,73 +1,122 @@
 
+### 3 . Setting VAPID Keys
+
+Run the following command to generate Public and Private VAPID :
```
I'd explain a bit what VAPID keys are. Suggestion:

----
VAPID keys are a public-private key pair used with Firebase Cloud Messaging (FCM) to securely identify the server sending web push notifications.

You can generate your own VAPID keys by running the following command:

technophile-04 replied:
Love this definition, to the point! 

Maybe : 
"VAPID keys are a public-private key pair used ~with Firebase Cloud Messaging (FCM)~ to securely identify the server sending web push notifications."

I think we can remove FCM from it since we are not using that service 🙌

Btw initially I was planning to use the Vercel KV store but was more comfortable with firestore as DB and looking at Vercel KV it felt hard for me to query from it (might be my very little exp with Redis)

 @carletex should we use Vercel KV since other BG app uses it or any other DB which is easy to configure?

carletex replied:
> I think we can remove FCM from it since we are not using that service 

My bad! ChatGPT tricked me haha. You definitions looks great.

> @carletex should we use Vercel KV since other BG app uses it or any other DB which is easy to configure?

We have a bunch of stuff using firebase (SRE, BGv3)... so that's why I didn't mind using it. We can start with Firebase and maybe we can tinker with KV later on.

Thanks!

carletex commented on README.md:
````diff
@@ -1,73 +1,122 @@
 ```
-yarn chain
+
+- Set `PUBLIC_KEY_VAPID` and `PRIVATE_KEY_VAPID` variables in `packages/nextjs/.env` file.
+- Set `NEXT_PUBLIC_PUBLIC_KEY_VAPID` in `packages/nextjs/.env.local` file.
````
A couple of questions here:

1. Do we need to use both files (.env.local / .env)? What's the issue with using only .env or .env.local for all?
2. Could we just have `NEXT_PUBLIC_PUBLIC_KEY_VAPID` and `PRIVATE_KEY_VAPID`?

technophile-04 replied:
Ohhh shit I had this wrong notion that `.env.local` are used to store only `NEXT_PUBLIC_*` variables (which are visible to client) and `.env` are used for non-NEXT_PUBLIC variables 🤦‍♂️

![Screenshot 2023-09-26 at 9 45 50 PM](https://github.com/BuidlGuidl/PWA-burner-wallet/assets/80153681/91844ecb-3fa7-455a-be5a-72aa727680c5)

This from Nextjs docs solved all my doubts and lol I was wondering at https://github.com/BuidlGuidl/buidlguidl.com/pull/17#issuecomment-1713413719 why don't we have `.env` added by default in `.gitignore` 

Nevertheless updated it at https://github.com/BuidlGuidl/PWA-burner-wallet/pull/1/commits/43e5b9acb94929128095266192facb3e43893c67 🙌

carletex replied:
Yep!

I think I never pushed an `.env` file to a Vercel project, because it feels wrong haha

carletex commented on README.md:
```diff
@@ -1,73 +1,122 @@
+3. Once its successful you will see _"Notify All"_ button => This button makes "POST" request to `packages/nextjs/pages/api/push/notify-all.ts` which will send notification to all the subscribers
+
+### 6 . Deploying your PWA to Vercel
+
```
Maybe adding the hint we have on SE-2?

> Hint: We recommend connecting your GitHub repo to Vercel (through the Vercel UI) so it gets automatically deployed when pushing to main.

carletex commented on README.md:
```diff
@@ -1,73 +1,122 @@
+
+> Hint: You can mass copy all the config variables from your`packages/nextjs/.env` and `packages/nextjs/.env.local` config files and paste them into the Vercel form.
+
+![VercelEnvironmentVariables](https://github.com/BuidlGuidl/PWA-burner-wallet/assets/55535804/8d1a56cf-a7c0-4ebe-8949-18673d6542e7)
```
This is the only screenshot we have in the README... I think we don't need it.

Pabl0cks replied:
Agree! I had the same feeling watching the lonely screenshot in the Readme, but my mind convinced me that we would have more screenshots soon so it would be fixed. But probably that screenshot is not even needed, and right now looks bad.

Deleted! 🙌

carletex commented on packages/nextjs/pages/api/push/add-subscription.ts:
```diff
@@ -0,0 +1,45 @@
+
+const PUBLIC_KEY_VAPID = process.env.PUBLIC_KEY_VAPID ?? "";
+const PRIVATE_KEY_VAPID = process.env.PRIVATE_KEY_VAPID ?? "";
+webpush.setVapidDetails("mailto:shivbhonde04@gmail.com", PUBLIC_KEY_VAPID, PRIVATE_KEY_VAPID);
```
Change it to `admin@buidlguidl.com` ?

carletex commented on packages/nextjs/.env.example:
```diff
@@ -1,3 +1,15 @@
+# PWA environment variables.
+# This all lives in the `.env` file, which is not committed to git
+PUBLIC_KEY_VAPID=
+PRIVATE_KEY_VAPID=
```
Maybe we can moved this after the Firebase ones... since we fill them last if we follow the README workflow.

technophile-04 replied:
fiexed at https://github.com/BuidlGuidl/PWA-burner-wallet/pull/1/commits/43e5b9acb94929128095266192facb3e43893c67 🙌

carletex commented on packages/nextjs/.env.example:
```diff
@@ -1,3 +1,15 @@
+# This all lives in the `.env` file, which is not committed to git
+PUBLIC_KEY_VAPID=
+PRIVATE_KEY_VAPID=
+FB_API_KEY=
+FB_AUTH_DOMAIN=
+FB_PROJECT_ID=
+FB_STRG_BUCKET=
+FB_MESSAGING_SENDER_ID=
+FB_APP_ID=
```
I keep seeing FACEBOOK when reading FB.

I think we could use `FIREBASE_`

technophile-04 replied:
lol yup makes sense ! updated at https://github.com/BuidlGuidl/PWA-burner-wallet/pull/1/commits/2add41ad38cd07fb2cb698a50c72c488ce82f054 🙌

carletex commented on README.md:
````diff
@@ -1,73 +1,122 @@
+```bash
+yarn web-push-generate
 ```
-yarn chain
````
I think we need chain & deploy, right?

If not you get an error when launching the app locally on hardhat.

Pabl0cks replied:
Just added the info to setup the local environment to the _"Starting the PWA"_ section.  

To keep it reduced, used the guide we use on SRE challenges, assuming users can read more details in SE-2 Docs, but not 100% sure if that assumption/approach is correct.

**I think could be interesting to add an extra section with the info on how to deploy to a real network**, even a reduced explanation could do the job, plus linking to the Docs sections if the user wants extra info (Deploy your Smart Contracts and Deploy your NextJS App sections). 

What do you think?

carletex replied:
Yep, I agree!

I think this is ready to merge. I'll set up a Vercel with auto-deployment + firebase for this and will merge to main. Probably next week, so feel free to add that section if you want @Pabl0cks (if not, it's not a big deal and we can iterate on a future PR)

Thanks!

Pabl0cks replied:
Section added! 🙌 
Was not sure about adding or not an initial sentence on why would you want to deploy to a public network, but I guess anyone building a PWA with SE-2 already knows why.

carletex commented:
Hey all! left some review comments.

Also, following the instructions, I get this when clicking "allow notification" (after accepting it in my browser/installed pwa)

![image](https://github.com/BuidlGuidl/PWA-burner-wallet/assets/2486142/25317fb0-d41d-49f0-a738-8967f4822c51)
![image](https://github.com/BuidlGuidl/PWA-burner-wallet/assets/2486142/d1805566-72bd-41fc-9e13-0135355b0c23)

Any ideas on what I might be doing wrong?

carletex commented on README.md:
````diff
@@ -1,73 +1,122 @@
 ```
 
-2. Run a local network in the first terminal:
+> Hint : Make sure you have read and write [rules](https://firebase.google.com/docs/firestore/security/get-started#testing_rules) allowed in Firebase.
````
I think we need to be more clear, that you need to set up a Firestore.

Firestore != Firebase, and creating a Firestore wasn't mentioned before.

We could also be more explicit / copy the Firestore rules here. Maybe only allow read/write for all in the `subscriptions` collection?

technophile-04 replied:
> I think we need to be more clear, that you need to set up a Firestore.

ohh yes we need to tell about [creating  firestore database](https://firebase.google.com/docs/firestore/quickstart)

> We could also be more explicit / copy the Firestore rules here. Maybe only allow read/write for all in the subscriptions collection?

Yup, @Pabl0cks please could you update this 🙌, here is the rule(tested it and works) : 

```
rules_version = '2';

service cloud.firestore {
  match /databases/{database}/documents {

    // This rule allows anyone with your Firestore database reference to view, edit,
    // and delete all data in your Firestore subscriptions collection
    match /subscriptions/{subscription} {
      allow read, write: if true;
    }
  }
}
```

Pabl0cks replied:
I've just pushed in https://github.com/BuidlGuidl/PWA-burner-wallet/pull/1/commits/4af299ac1d8abcd727f7524e546a89b49cb5d9f1 the updated version for **2 . Setting up Firebase** section. 

Let me know if you would change anything else there ♥

carletex replied:
@technophile-04 I saw you added this:

```
// This rule allows anyone with your Firestore database reference to view, edit,
// and delete all data in your Firestore subscriptions collection
```

Maybe we can move that warning out of the code block? It seems that it's not the ideal setup (for security reasons)

I don't understand the full picture yet: with that rule, who is able to write? Everyone in the front-end? Or just from a backed that have the secret? Maybe it's just because we have the "notify all" option in the testing app.

We could also think a better example. In any case, I think it's important to warn people about that setting.

thanks!

technophile-04 replied:
> I don't understand the full picture yet: with that rule, who is able to write? Everyone in the front-end? Or just from a backed that have the secret? Maybe it's just because we have the "notify all" option in the testing app.

I actually copied and pasted those lines from when you selected Firebase test mode, but after doing some research 

It seems that if anyone has your "project-id" and "webiapi-key" then can just make API request to firestore REST API : 

```
curl -X DELETET https://firestore.googleapis.com/v1/projects/{PROJECT_ID}/databases/(default)/documents/subscriptions/{DOCUMENT_ID}?key={web-api-key}
```

But this is not the case with us since we not are not leaking PROJECT_ID, WEB_API_KEY because we are accessing Firebase through NEXTJs serverless functions 

If we would have used Firebase on client side stored those variable in `NEXT_PUBLIC` then people can get those keys and make a malicious request if the rules were not set properly

carletex replied:
> because we are accessing Firebase through NEXTJs serverless functions

Great!! Thank you :)

technophile-04 committed:
update mail

technophile-04 committed:
use .env.local to store env variables and use NEXT_PUBLIC_PUBLIC_KEY_VAPID

technophile-04 committed:
rename FB => FIREBASE

technophile-04 committed:
add connect to github hint in deploying you app section

Pabl0cks committed:
Add environment setup info to Starting the PWA section

Pabl0cks committed:
Add VAPID keys explanation

Pabl0cks committed:
Delete Vercel Environment Variables screenshot

Pabl0cks committed:
Simplify Firebase credentials explanation and add Firestore database creation and rules guide

Pabl0cks committed:
Add section to explain how to deploy contracts to public networks

carletex committed:
Typo

carletex commented:
I think this is ready for a first merge.

I'm gonna create a branch "deployed-demo", where I'll branch from `main` and will use a Sepolia Contract so it works... until we firgure out next steps. It'll be connected to the BG's Vercel (something like pwa-demo.buidlguidl.com)

Thanks @technophile-04 & @Pabl0cks !
